scvi/train/_trainingplans.py

```diff
@@ -708,6 +709,37 @@ def __init__(
         )
         self.loss_kwargs.update({"classification_ratio": classification_ratio})
 
+    def compute_and_log_metrics(
+        self, loss_output: LossOutput, metrics: Dict[str, ElboMetric], mode: str
+    ):
+        """Computes and logs metrics."""
+        super().compute_and_log_metrics(loss_output, metrics, mode)
+
+        # no labeled observations in minibatch
+        if loss_output.classification == {}:
+            return
+
+        ce_loss = loss_output.classification[METRIC_KEYS.CROSS_ENTROPY_KEY]
+        true_labels = loss_output.classification[METRIC_KEYS.TRUE_LABELS_KEY]
+        logits = loss_output.classification[METRIC_KEYS.LOGITS_KEY]
+        predicted_labels = torch.argmax(logits, dim=-1, keepdim=True)
+        n_classes = loss_output.classification[METRIC_KEYS.N_CLASSES_KEY]
+
+        accuracy = Accuracy(task="multiclass", num_classes=n_classes).to(logits.device)(
+            predicted_labels, true_labels
+        )
+        f1 = F1Score(task="multiclass", num_classes=n_classes).to(logits.device)(
+            predicted_labels, true_labels
+        )
+        auroc = AUROC(task="multiclass", num_classes=n_classes).to(logits.device)(
+            logits, true_labels.view(-1).long()
+        )
```

Comment:
is there a way to avoid `.to`? In general this can be problematic on multi device training. I see it might be because you don't know n_classes a priori.

Reply:
I think it should be possible by moving the tensors to CPU before logging and then the metric will be on the same device by default

Reply:
I though if you decalre the metrics in the init, lightning will handle all of this? Maybe we could detect n_classes in a different way

Reply:
Makes sense. n_classes can be passed into the training plan init from scanvi